src/main/java/org/folio/innreach/batch/contribution/service/InitialContributionEventProcessor.java

```diff
@@ -0,0 +1,178 @@
+import static org.folio.innreach.domain.entity.JobExecutionStatus.Status.PROCESSED;
+import static org.folio.innreach.domain.entity.JobExecutionStatus.Status.READY;
+import static org.folio.innreach.domain.entity.JobExecutionStatus.Status.RETRY;
+
+@Service
+@Log4j2
+@RequiredArgsConstructor
+public class InitialContributionEventProcessor {
+  @Qualifier("instanceExceptionListener")
+  private final ContributionExceptionListener instanceExceptionListener;
+  @Qualifier("itemExceptionListener")
+  private final ContributionExceptionListener itemExceptionListener;
+  private final InventoryViewService inventoryViewService;
+  private final RecordContributionService recordContributionService;
+  private final ContributionValidationService validationService;
+  private final ContributionJobProperties jobProperties;
+  private final JobExecutionStatusRepository jobExecutionStatusRepository;
+  private static final ConcurrentHashMap<UUID, Contribution> contributionRecord = new ConcurrentHashMap<>();
+  private final ContributionRepository contributionRepository;
+  private final TenantScopedExecutionService executionService;
+  @Value("${initial-contribution.retry-attempts}")
+  private int maxRetryAttempts;
+
+  @Transactional
+  @Async
+  public void processInitialContributionEvents(JobExecutionStatus job) {
+    executionService.runTenantScoped(job.getTenant(), () -> {
+      log.info("processInitialContributionEvents:: Processing Initial contribution events {} ", job);
+      try {
+        var instanceId = job.getInstanceId();
+        var centralServerId = contributionRecord.get(job.getJobId()) != null ?
+          contributionRecord.get(job.getJobId()).getCentralServer().getId() : getCentralServerId(job.getJobId());
+        var instance = inventoryViewService.getInstance(instanceId);
+        if (centralServerId == null || instance == null) {
+          log.warn("processInitialContributionEvents:: Unable to process event with instance " +
+            "id {} centralServerId {} ", instanceId, centralServerId);
+          updateJobAndContributionStatus(job, FAILED, job.isInstanceContributed());
+          return;
+        }
+        checkRetryLimit(job);
+        startContribution(centralServerId, instance, job);
+      } catch (ServiceSuspendedException | InnReachConnectionException | FeignException |
+               SocketTimeOutExceptionWrapper ex) {
+        log.warn("processInitialContributionEvents:: Retrying the contribution for instanceId {} due to {} ",
+          job.getInstanceId(), ex.getMessage());
+        updateJobAndContributionStatus(job, RETRY, job.isInstanceContributed());
+      } catch (Exception ex) {
+        log.warn("processInitialContributionEvents:: Exception while processing instanceId {} ", job.getInstanceId());
+        logException(job, ex, contributionRecord.get(job.getJobId()).getId());
+        updateJobAndContributionStatus(job, FAILED, job.isInstanceContributed());
+      }
+    });
+  }
```

Comment:
I'm trying to imagine what it will be like keeping tabs on this job while it is running. I know we're wanting to remove the statistics because the design in the `ContributionJobRunner` was way too tied to the application (not microservices).

And yet we still have a need to keep statistics of some kind, even if it is just log messages telling us how many messages have been processed. **How else will we know that things are working?** I think there is also an UI that relies on the statistics table which the end user uses to get this status.

How can we solve for this?

One idea I know we have talked about is to use the outbox table as the source of truth. We could query this while the job runs, running a `count` on outbox the table, for various statuses, perhaps when updating in `updateJobAndContributionStatus`, and then update the statistics table in the DB with the current counts of each state. Because as it stands now I think the status will always show 0 for the job unless I'm missing something.

We already have `ContributionJobStatsListener.updateStats` which writes the stats to storage. We could just populate this with stats that are the result of our `count` sql against the job table. We could also log this information to an `info` statement so we can keep tabs on it in CW.

Reply:
I didn't notice that the stats were being upgraded elsewhere.